src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/CoreCLRReadyToRunHeaderNode.cs

```diff
@@ -0,0 +1,119 @@
+                Id = id;
+                Node = node;
+                StartSymbol = startSymbol;
+            }
+
+            readonly public ReadyToRunSectionType Id;
+            readonly public ObjectNode Node;
+            readonly public ISymbolNode StartSymbol;
+        }
+
+        List<HeaderItem> _items = new List<HeaderItem>();
+        TargetDetails _target;
+
+        public CoreCLRReadyToRunHeaderNode(TargetDetails target)
+        {
+            _target = target;
+        }
+
+        public void Add(ReadyToRunSectionType id, ObjectNode node, ISymbolNode startSymbol)
+        {
+            _items.Add(new HeaderItem(id, node, startSymbol));
+        }
+
+        public void AppendMangledName(NameMangler nameMangler, Utf8StringBuilder sb)
+        {
+            sb.Append(nameMangler.CompilationUnitPrefix);
+            sb.Append("__ReadyToRunHeader");
+        }
+        public int Offset => 0;
+        public override bool IsShareable => false;
+
+        protected override string GetName(NodeFactory factory) => this.GetMangledName(factory.NameMangler);
+
+        public override bool StaticDependenciesAreComputed => true;
+
+        public override ObjectNodeSection Section
+        {
+            get
+            {
+                if (_target.IsWindows)
+                    return ObjectNodeSection.ReadOnlyDataSection;
+                else
+                    return ObjectNodeSection.DataSection;
+            }
+        }
+
+        public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
+        {
+            ObjectDataBuilder builder = new ObjectDataBuilder(factory, relocsOnly);
+            builder.RequireInitialPointerAlignment();
+            builder.AddSymbol(this);
+
+            // Don't bother sorting if we're not emitting the contents
+            if (!relocsOnly)
+                _items.Sort((x, y) => Comparer<int>.Default.Compare((int)x.Id, (int)y.Id));
+
+            // ReadyToRunHeader.Magic
+            builder.EmitInt((int)(ReadyToRunHeaderConstants.Signature));
+
+            // ReadyToRunHeader.MajorVersion
+            builder.EmitShort((short)(ReadyToRunHeaderConstants.CurrentMajorVersion));
+            builder.EmitShort((short)(ReadyToRunHeaderConstants.CurrentMinorVersion));
+
+            // ReadyToRunHeader.Flags
+            builder.EmitInt(0);
+
+            // ReadyToRunHeader.NumberOfSections
+            ObjectDataBuilder.Reservation sectionCountReservation = builder.ReserveInt();
+            
+            int count = 0;
+            foreach (var item in _items)
+            {
+                // Skip empty entries
+                if (!relocsOnly && item.Node.ShouldSkipEmittingObjectNode(factory))
+                    continue;
+
+                builder.EmitInt((int)item.Id);
+                
+                builder.EmitReloc(item.StartSymbol, RelocType.IMAGE_REL_BASED_ADDR32NB);
+                builder.EmitInt(item.Node.GetData(factory).Data.Length);
```

Comment:
Might be good to put the `GetData` call under `if (!relocsOnly)`. Depending on what nodes the header is going to refer to, some of them might become frozen after a `GetData` call, and it might be a pretty expensive call.

Reply:
Agreed - changed